### PR TITLE
[WIP] Adding Instagram Filters to Image Viewer via Caman.js (fixes issue #492)

### PR DESCRIPTION
**Pull request by timmoy**
# Initial Commit Summary
I've created a new file called "instagram.js" in the "src/thirdparty/caman" directory that contains code similar to my [implementation for the selfie taker](https://github.com/mozilla/brackets/pull/656). 
The file is in the thirdparty folder to try and avoid the linting error from the previous pull pull request. This seems to have solved the problem for the time being.

We're also using a separate canvas because the Caman function to replace the  element with a canvas won't execute synchronously. So the variable "canvas" is assigned a HtmlImageElement and canvas.getContaxt('2d') can't be invoked properly.

The savePhoto() function was ported via some mix of module importing and hard coding some stuff that couldn't be imported (since we don't use a camera object).

The image-view.html file was also modified to include the basic UI for the filters and script tags to load the library and the new file.

# Initial Commit To Do Breakdown
## Bugs to be fixed (shown in the gif in the order described):
- loss of swatches due to script tags being added
- can't save the filtered image to replace the original
- the canvas also disappears if you open a new image (basically loads for the first image opened per session; if you refresh the page it will lose the canvas for the first image only)
caman initial image viewer bugs

## Currently Working:
- Loading from image to canvas
- Filtering and reset buttons

## Other Improvements to be made:
- Single canvas loading and filtering
- UI also needs to be improved, but that is the focus after the bugs are fixed.

### CODE REVIEW

File: src/thirdparty/caman/instagram.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  var fs = brackets.getModule("fileSystemImpl");
+  var Buffer = brackets.getModule("filesystem/impls/filer/BracketsFiler").Buffer;
+
+  function savePhoto(){
+    Caman(canvas, function () {
+      this.render(function () {
+            var image = this.toBase64();
+            var binaryDataStr = /^data:image\/png;base64,(.+)/.exec(image)[1];
+            //self.camera.savePhoto(base64ToBuffer(binaryDataStr));
+              var binary = window.atob(binaryDataStr);
```
Comment (humphd) [inline src/thirdparty/caman/instagram.js]: Not all browsers support `window.atob`, specifically I don't think IE has it.  We might already have this code somewhere in Brackets, and can expose it.  We should figure this out.

File: src/thirdparty/caman/instagram.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+              var len = binary.length;
+              var bytes = new Uint8Array(len);
+              for(var i = 0; i < len; i++) {
+                bytes[i] = binary.charCodeAt(i);
+              }
+              var data = new Buffer(bytes.buffer);
+              fs.writeFile(savePath, data, {encoding: null}, function() {
```
Comment (humphd) [inline src/thirdparty/caman/instagram.js]: This needs error handling in the callback.

File: src/thirdparty/caman/instagram.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                console.log("save path: " + savePath);
+              });
+           console.log("attributes: " + canvas.width + ", " + canvas.height);
+         });
+       });
+       console.log("photo saved");
+  }
+
+  /*
+  Can't use this pattern at the moment since the Caman function won't run before
```
Comment (humphd) [inline src/thirdparty/caman/instagram.js]: It means you need to add an event for the DOM to be loaded, maybe use `DOMContentLoaded`

File: src/thirdparty/caman/instagram.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  //img.src = canvas.src;
+  img.src = document.getElementById('image-canvas').src;
+  var savePath = img.src;
+  console.log("AFTER src = " + img.src + " from: " + canvas.src);
+
+  img.onload = function() {
+    canvas.width = img.width;
+    canvas.height = img.height;
+    ctx.drawImage(img, 0, 0, img.width, img.height);
+  };
```
Comment (humphd) [inline src/thirdparty/caman/instagram.js]: You probably need an `img.onerror` too, for cases where things go wrong.

**Comment by flukeout** [comment]
@timmoy - let me know once there is some markup to work with, look at or style and I'll be happy to jump in.

File: src/thirdparty/caman/instagram.js (hunk 1 above)
Comment (timmoy) [inline src/thirdparty/caman/instagram.js]: I did a search on its compatibility and it seems that IE 10 and onward can use it. Does this mean we should find a workaround for the method for versions before IE 10?

Comment (humphd) [inline src/thirdparty/caman/instagram.js]: We support:

* IE: 10+ (IndexedDB)
* Firefox: 26+ (IndexedDB)
* Chrome: 31+ (IndexedDB, WebSQL)
* Safari: 7.0+ (WebSQL)
* Opera: 19+ (IndexedDB, WebSQL)

**Comment by timmoy** [comment]
@humphd I got an interesting result when using DOMContentLoaded: the event never occurred. I think it might be the script tags causing the issue as they seem to also be breaking the colour swatch feature. Though I'm at a loss as to why this problem exists for the image viewer but not the selfie-taker.

Anyways, the current workaround I'm using is to reduce the img to 0 pixels and let the image-viewer do its thing to get the image url via {{imgUrl}} that we use to load the image into the canvas. Using {{imgUrl}} caused the file to be undefined, so I had to keep the  object around in some capacity for getting the url.

**Comment by timmoy** [comment]
So, with the new commit I've been able to solve two of the bugs at the expense of putting things back into the ImageViewer.js file (so we're failing the lint check again). 

But, this means I was able to remove the script tags for the html file which led me to putting the filtering functionality into the onImageLoaded function, which let's us load the Caman.js stuff every time we click on a new image. Also, since the script tags are gone I have been able to get the swatches to show again. 

Below is a gif demonstrating the current commit (note saving still doesn't work).

caman initial image viewer update

I have 2 images since the swatches extension extracts the colours from the img element so removing it causes it to throw an exception and "hide". 
Replacing the img element after the extractColors function could be possible, so I'll test that out in the next few days along with trying to find out the exact problem of why saving can't work.

The code is a huge mess right now, but I'll have it cleaned up after things settle down.

@humphd @gideonthomas @Pomax Let me know if you think I'm heading in the wrong direction or should be making adjustments.

@flukeout I think it's possible we can start on a bit of the design for the interface. I'm using tables and some basic divs and buttons like before in the "image-view.html" [changes here](https://github.com/timmoy/brackets/blob/a0f1ee645f25fd055a117b82eb3cd947a03d76fe/src/htmlContent/image-view.html#L32-L87).

**Comment by flukeout** [comment]
@timmoy - is there an email I can reach you at? I'm collecting information for a contributors page on the Thimble website and I'd like to include you. Can you please email me at ``luke@mozillafoundation.org``? Thanks!

**Comment by flukeout** [comment]
Also, I'll take a look at the styling and markup probably sometime tomorrow. Most likely, I'll make a pull request against your branch.

**Comment by gideonthomas** [comment]
@timmoy thanks for your work on this! I'm going to take what you've done so far, modify it a bit and open up a PR against your branch since the changes that I'm suggesting are a bit complex. Once that happens, let me know if it makes sense and feel free to suggest changes as well.

**Comment by gideonthomas** [comment]
@timmoy I've made a PR against your branch: https://github.com/timmoy/brackets/pull/2. See if it makes sense to you (if not, feel free to ask questions) and then merge it in. Then, @flukeout can suggest styling changes.

**Comment by timmoy** [comment]
@gideonthomas I merged the branch and github prompted me to fix some merge conflicts. I also brought the changes into my local machine and found a lot of commits and files changed. This means that I shouldn't need to rebase onto master, right?

**Comment by gideonthomas** [comment]
@timmoy you definitely need to rebase onto master because my patch was only on top of yours and you need the commits from master as well.

**Comment by timmoy** [comment]
@gideonthomas I've rebased the code as requested and it should be up to date with master, plus the seven commits in this branch. I suppose that we can go ahead with @flukeout for the styling changes needed?

**Comment by timmoy** [comment]
@gideonthomas I've noticed after my rebase that button text has disappeared for the most part (selfie taker buttons still work). See the gif below.
caman some post rebase bugs
It seems that after the rebase, the moustache code you did for the strings isn't loading. Is this just for my system and should be ignored?

Functionality is working fine although with the last commit I did, I changed [this line](https://github.com/gideonthomas/brackets/blob/ce677ca4168a23f90a6fce23c7493f6aa16ecdb5/src/editor/ImageViewer.js#L162) back to your version. When using the current version that master does like [this](var dimensionString = this._naturalWidth + " &times; " + this._naturalHeight + " " + Strings.UNIT_PIXELS;), we get a "str" is undefined somehow... I'm not sure how to proceed here since we're kind of regressing this part of the code?

File: src/editor/ImageViewer.js [hunk 5]
```diff
@@ -149,8 +155,7 @@ define(function (require, exports, module) {
 
         var extension = FileUtils.getFileExtension(this.file.fullPath);
 
-        var stringFormat = Strings.IMAGE_DIMENSIONS;
-        var dimensionString = StringUtils.format(stringFormat, this._naturalWidth, this._naturalHeight);
+        var dimensionString = this._naturalWidth + " &times; " + this._naturalHeight + " " + Strings.UNIT_PIXELS;
```
Comment (gideonthomas) [inline src/editor/ImageViewer.js]: please undo this change

**Comment by timmoy** [comment]
@gideonthomas I've reverted the change, and can confirm it's working with swatches, strings and save functionality still intact.

File: src/editor/ImageViewer.js [hunk 6]
```diff
@@ -36,7 +36,8 @@ define(function (require, exports, module) {
         BlobUtils           = require("filesystem/impls/filer/BlobUtils"),
         FileUtils           = require("file/FileUtils"),
         _                   = require("thirdparty/lodash"),
-        Mustache            = require("thirdparty/mustache/mustache");
+        Mustache            = require("thirdparty/mustache/mustache"),
+        Image               = require("editor/image");
```
Comment (gideonthomas) [inline src/editor/ImageViewer.js]: needs to be capital `I`

Comment (timmoy) [inline src/editor/ImageViewer.js]: I've fixed the issue; thanks for catching that.

File: src/filesystem/impls/filer/FilerFileSystem.js [hunk 7]
```diff
@@ -429,6 +430,19 @@ define(function (require, exports, module) {
         callback();
     }
 
+    // Based on http://stackoverflow.com/questions/21797299/convert-base64-string-to-arraybuffer
```
Comment (humphd) [inline src/filesystem/impls/filer/FilerFileSystem.js]: Why not put this in [`FilerUtils.js`](https://github.com/timmoy/brackets/blob/7c744eb5675c7243d749c62a26c4cf07d9c80b41/src/filesystem/impls/filer/FilerUtils.js), since you're not using it in here at all?

File: src/main.js [hunk 8]
```diff
@@ -37,6 +37,7 @@ require.config({
         // implementations (e.g. cloud-based storage).
         "fileSystemImpl"    : "filesystem/impls/filer/FilerFileSystem",
 
+        "caman"             : "thirdparty/caman/caman.full.min",
```
Comment (humphd) [inline src/main.js]: This is probably the right thing to do, but let me play devil's advocate: why not move this into a `<script>` block in `src/htmlContent/image-view.html` so that it only loads if/when we need it vs. always.  It's 76K of code we're loading for everyone, when only some people will use it.

Comment (timmoy) [inline src/main.js]: The main reason for not using `<script>` blocks was that that would only load the script once each browser session. Basically, the image viewer would be loaded once per browser session, and as such so would our script. This caused major issues when there were multiple images as the script wouldn't run again to update the image in the canvas, thus leading to blank canvases for all images once a new image was viewed. Since CamanJS won't work without a canvas, we needed to move the code to [this function](https://github.com/timmoy/brackets/blob/7c744eb5675c7243d749c62a26c4cf07d9c80b41/src/editor/ImageViewer.js#L151-L188) so that everytime the image loads we'll create a canvas for that image. There were also side effects of breaking the colour swatch extension when using the `<script>` blocks, but I'm not sure why that was.

Also from the same bug, using the `<script>` blocks to load the library ended up causing issues with the way that the editor was loaded. I remember that it was an issue with `DOMContentLoaded` and using a CamanJS function to replace the image with a canvas that showed the `DOMContentLoaded` event never fired as well as noting that asynchronous loading was being used after seeing the order of my debug statements and how the CamanJS function tried to replace a null object. I experimented with trying to force it to load the function synchronously via callbacks, but failed as brackets seems to be able to disable it somehow?

After doing some searching on the full explanation, I found someone with a [similar bug](https://bugzilla.mozilla.org/show_bug.cgi?id=667227#c8) on bugzilla that dealt with iframes and `<script>` blocks failing to load. It's not too related to the bug that happened here, but can be a valuable resource.

File: src/filesystem/impls/filer/FilerFileSystem.js (hunk 7 above)
Comment (timmoy) [inline src/filesystem/impls/filer/FilerFileSystem.js]: True, it does seem a bit odd. I did some digging into the code and the quick explanation is that there are other functions in `FilerFileSystem.js` that we use that aren't in `FilerUtils.js`. So if we put it in `FilerUtils.js`, then we'll need two requires or getModule statements that we're accomplishing with one instead right now.

Extra:
We actually use [writeFile](https://github.com/timmoy/brackets/blob/7c744eb5675c7243d749c62a26c4cf07d9c80b41/src/filesystem/impls/filer/FilerFileSystem.js#L257) in both of the save functions for the [camera](https://github.com/timmoy/brackets/blob/7c744eb5675c7243d749c62a26c4cf07d9c80b41/src/extensions/default/BrambleUrlCodeHints/camera/index.js#L49) and [image viewer](https://github.com/timmoy/brackets/blob/7c744eb5675c7243d749c62a26c4cf07d9c80b41/src/editor/Image.js#L26). Since we need both the base64ToBuffer function and writeFile function for saving, it reduces the overhead of pulling files by including the function we ported in another file that has code we need to work with. It is also quite relevant to the FileSystem, so it's placement still makes sense unlike if it was placed into a non file system file.

**Comment by timmoy** [comment]
@humphd I've update the caman file with the comment block. Please review and let me know if I need to add/remove any information.

File: src/filesystem/impls/filer/FilerFileSystem.js (hunk 7 above)
Comment (gideonthomas) [inline src/filesystem/impls/filer/FilerFileSystem.js]: @timmoy there won't be any overhead because requires are cached. I agree with @humphd that this probably should be in FilerUtils.js instead. We can fix other issues with functions that aren't in there right now that should be in FilerUtils later.

File: src/main.js (hunk 8 above)
Comment (gideonthomas) [inline src/main.js]: @humphd we would need to deal with load timing issues in that case. One thing I would like to do instead (in a follow-up issue) is to remove a bunch of the cruft in Caman.js that we don't use, i.e. create a customized built file and use that instead. I bet we can reduce the size significantly with that.

File: src/filesystem/impls/filer/FilerFileSystem.js (hunk 7 above)
Comment (timmoy) [inline src/filesystem/impls/filer/FilerFileSystem.js]: @gideonthomas that's quite handy that the requires are cached; I've gone ahead and made a commit to port over the code from FilerFileSystem. I tested and it works for the selfie taker and image viewer. Can you review in case I missed anything?

File: src/editor/ImageViewer.js (hunk 5 above)
Comment (timmoy) [inline src/editor/ImageViewer.js]: the change has been undone as requested

File: src/editor/Image.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
+define(function (require, exports, module) {
+    "use strict";
+
+    var Caman           = require("caman");
+    var FilerFileSystem = require("fileSystemImpl");
+    var FileUtils       = require("filesystem/impls/filer/FilerUtils");
```
Comment (gideonthomas) [inline src/editor/Image.js]: FilerUtils vs. FileUtils. Missing 'r'

**Comment by gideonthomas** [comment]
One small nit change (FilerUtils vs. FileUtils) that needs to be made, r+ from me otherwise. Really nice work on this @timmoy!

Will leave the final review on @humphd since I contributed to some of this code.

File: src/editor/Image.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            var imageBase64Data = image.canvas.toDataURL(imageMimeType);
+            var data = FilerUtils.base64ToBuffer(imageDataRegex.exec(imageBase64Data)[1]);
+
+            FilerFileSystem.writeFile(imagePath, data, {encoding: null}, function(err) {
+                if(err) {
+                    console.error("[Bramble] Image with filters failed to save with: ", err);
+                    return;
+                }
+
+                FileSystemCache.refresh(function(err) {
```
Comment (humphd) [inline src/editor/Image.js]: Do you really need this?  It seems unnecessary to me, since you're always going to be overwriting the original file.  It won't hurt anything, but it's wasteful.  Maybe file a bug later to rip it out if I'm right, and it's not needed.

File: locales/en-US/editor.properties [hunk 11]
```diff
@@ -140,6 +140,19 @@ DND_SUCCESS_UNTAR=Successfully untarred <b>{0}</b>.
 # Image Viewer
 IMAGE_DIMENSIONS={0} (width) &times; {1} (height) pixels
 
+# Image Filters
+
+IMAGE_FILTER=Add Filter
```
Comment (humphd) [inline locales/en-US/editor.properties]: `IMAGE_FILTERS_TITLE` would make more sense to me, but it's not really important.

File: src/editor/Image.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
+define(function (require, exports, module) {
+    "use strict";
+
+    var Caman           = require("caman");
+    var FilerFileSystem = require("fileSystemImpl");
+    var FilerUtils       = require("filesystem/impls/filer/FilerUtils");
```
Comment (humphd) [inline src/editor/Image.js]: nit: alignment here.

File: src/editor/Image.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+                    $resetBtn.prop("disabled", true);
+                });
+            });
+        });
+
+        function applyFilterFn(fnName, args) {
+            image.reset();
+            image[fnName].apply(image, args);
+            image.render();
+            $saveBtn.prop("disabled", false);
```
Comment (humphd) [inline src/editor/Image.js]: Should you disable the filter buttons at the top of this function so that people don't click it over and over while it's running?  It seems to do odd things for me if I click it multiple times quickly.  Again, could be done in follow-up bug.

File: src/extensions/default/BrambleUrlCodeHints/camera/index.js [hunk 14]
```diff
@@ -8,11 +8,11 @@ define(function (require, exports, module) {
 
     var CommandManager = brackets.getModule("command/CommandManager");
     var Commands       = brackets.getModule("command/Commands");
+    var fs             = brackets.getModule("fileSystemImpl");
```
Comment (humphd) [inline src/extensions/default/BrambleUrlCodeHints/camera/index.js]: Curious why you chose to use this level of abstraction for the filesystem?  We don't do caching at this level.  Is this why you're doing the `FileSystemCache.refresh`?

File: src/thirdparty/caman/caman.full.min.js [hunk 15]
```diff
@@ -0,0 +1,244 @@
+/*
```
Comment (humphd) [inline src/thirdparty/caman/caman.full.min.js]: How come you didn't just use `npm` to install this vs. pulling it directly into the tree?  Not a blocker, but seems like something you could have done--other parts of Brackets do this now.

**Comment by humphd** [comment]
I think this is probably good to go.  Two things I think we could do in follow-up bugs:

screen shot 2017-05-07 at 10 02 26 pm

When we apply the filter, and the image's colours change, we should probably re-run the palette extraction code (e.g., on `Save` or something) so it properly reflects what's actually there vs. the original.

screen shot 2017-05-07 at 10 01 32 pm

It would be nice to separate the filter buttons from the Save/Reset buttons a bit more.  Did @flukeout do any UI review on this?  Can be something we iterate on.

Great work on this, it's very cool!

File: src/thirdparty/caman/caman.full.min.js (hunk 15 above)
Comment (gideonthomas) [inline src/thirdparty/caman/caman.full.min.js]: Interesting! Yeah I can try to do that.

**Comment by gideonthomas** [comment]
this is pending @flukeout's UI review

File: src/extensions/default/BrambleUrlCodeHints/camera/index.js (hunk 14 above)
Comment (gideonthomas) [inline src/extensions/default/BrambleUrlCodeHints/camera/index.js]: yeah basically! We seem to be using this in most of our bramble extensions so thought we should use it here too. Is there something else we should use that's better?

**Comment by flukeout** [comment]
Okay, checking it out now.

**Comment by flukeout** [comment]
In testing this, I'm unable to actually apply the filter. Here's a screencap gif of my process...

_caman

Console is wigging out...

image

cc @gideonthomas am I missing some steps in how I'm testing this other than just running the code in the PR?

**Comment by flukeout** [comment]
Okay, I submitted a PR against this branch that has my changes - https://github.com/timmoy/brackets/pull/3

**Comment by flukeout** [comment]
Okay, I figured out that the reason the filter wasn't working for me in the GIF above is because I'm using the image as a background-image. It looks like Thimble is setting the background image in the preview as a blob, which somehow gets corrupted or broken when the filter is applied...

image

cc @humphd @gideonthomas it seems to work fine for ```` elements however.

**Comment by humphd** [comment]
@flukeout please put the entire project you're doing online so I can see it.  It has to be something specific to the project (e.g., rewriting the image multiple times) or something.  I'll try to debug.

File: src/editor/Image.js (hunk 13 above)
Comment (timmoy) [inline src/editor/Image.js]: I agree, it does make the user experience a bit weird. I believe @flukeout has addressed it in his pull request as part of the UI changes. I'll probably merge it in sometime tomorrow evening.

File: src/editor/Image.js (hunk 10 above)
Comment (timmoy) [inline src/editor/Image.js]: I tried removing the function (but kept the button disabling) and it seems that it works without it on my machine. I'm not sure if it'll impact other sections of the code. 

@gideonthomas can you confirm if removing it was okay or that it's needed for some other part of the logic?

File: src/editor/Image.js (hunk 13 above)
Comment (humphd) [inline src/editor/Image.js]: @timmoy we're trying to freeze all the new code on our staging box by Thurs, so it's key that we get this stuff landed ASAP, so we can start testing.  If you can't get this done, let us know, and we'll take this over.

File: src/thirdparty/caman/caman.full.min.js (hunk 15 above)
Comment (humphd) [inline src/thirdparty/caman/caman.full.min.js]: It's fine if this happens in a follow-up, it doesn't need to block this from landing.

File: src/editor/Image.js (hunk 10 above)
Comment (humphd) [inline src/editor/Image.js]: I'll debug this a bit today as I try to figure out @flukeout's issue with `background-image` not updating in the preview.

File: src/editor/Image.js (hunk 13 above)
Comment (timmoy) [inline src/editor/Image.js]: @humphd I've taken most of the solution (no spinner) from @flukeout 's [pull request](https://github.com/timmoy/brackets/pull/3/commits/71871c85abc1092471aecbfdd31efa684be14084). 

I've tested it and it works well on my machine. I think this can be (or at least pretty close to being) landed?

**Comment by humphd** [comment]
I'm going to defer to @gideonthomas, who I think has been working on this too, today.

**Comment by gideonthomas** [comment]
Moved to https://github.com/mozilla/brackets/pull/756
